File: pascal/humanAB.pas
{$mode Delphi}

Program humanAB;

{
    I am useing Free Pascal to test.
}

Uses 
Sysutils, classes;

Function GetA(Guess: String; Answer: String): Integer;

Var 
  index1: Integer;
  length1: LongInt;
  length2: LongInt;
Begin
  length1 := Length(Guess);
  length2 := Length(Answer);
  If (length1 <> 4) Or (length2 <> 4) Then
    Begin
      Exit(0);
    End;

  Result := 0;
  For index1 := 1 To 4 Do
    Begin
      If (Guess[index1] = Answer[index1]) Then
        Begin
          Result := Result + 1;
        End;
    End;
End;

Function GetB(Guess: String; Answer: String): Integer;

Var 
  index1: Integer;
  index2: Integer;
  length1: LongInt;
  length2: LongInt;
Begin
  length1 := Length(Guess);
  length2 := Length(Answer);
  If (length1 <> 4) Or (length2 <> 4) Then
    Begin
      Exit(0);
    End;

  Result := 0;
  For index1 := 1 To 4 Do
    Begin
      For index2 := 1 To 4 Do
        Begin
          If (index1 <> index2) Then
            Begin
              If (Guess[index1] = Answer[index2]) Then
                Begin
                  Result := Result + 1;
                End;
            End;
        End;
    End;
End;

Var 
  stringlist : TStringList;
  RNumber : Longint;
  Guess   : String;
  Answer  : String;
  i, j, k, m : Integer;
  n: extended;
  aresult: Integer;
  bresult: Integer;

Begin
(* Initialize random number generator *)
  Randomize;
  stringlist := TStringList.Create;

  Try

    For i := 0 To 9 Do
      Begin
        For j := 0 To 9 Do
          Begin
            For k := 0 To 9 Do
              Begin
                For m := 0 To 9 Do
                  Begin
                    If (i <> j) And (i <> k) And (i <> m) And
                       (j <> k) And (j <> m) And (k <> m) Then
                      Begin
                        stringlist.add(IntToStr(i) + IntToStr(j) + IntToStr(k) + IntToStr(m));
                      End;
                  End;
              End;
          End;
      End;

     (* Pick up a number to be the answer *)
    RNumber := Random (stringlist.Count);
    Answer := stringlist[RNumber];

    While true Do
      Begin
        write('Please input an answer: ');
        readln(Guess);

        (* Try to check this string is numeric *)
        If Not TryStrToFloat(Guess, n) Then
          Begin
            writeln('It is not a valid input.');
            writeln();
            continue;
          End;

        aresult := GetA(Guess, Answer);
        bresult := GetB(Guess, Answer);
        If (aresult = 4) And (bresult = 0) Then
          Begin
            writeln('You guess the correct answer.');
            break;
          End;

        writeln('The result: ', aresult, ' A, ', bresult, ' B.');
        writeln();
      End;

  Finally
    FreeAndNIL(stringlist);
End;
End.

File: c++/humanAB.cpp
/*
 * g++ -std=c++11 humanAB.cpp -o humanAB
 */

#include <cstdio>
#include <cstdlib>
#include <ctime>
#include <iostream>
#include <string>
#include <vector>


int getA(std::string &answer, std::string &guess)
{
    int a = 0;

    if(answer.length() != guess.length()) {
        return 0;
    }

    for(int index = 0; index < answer.length(); index++) {
      if(answer[index] == guess[index]) {
        a++;
      }
    }

    return a;
}


int getB(std::string &answer, std::string &guess)
{
    int b = 0;

    if(answer.length() != guess.length()) {
        return 0;
    }

    for(int index = 0; index < answer.length(); index++) {
      for(int index2 = 0; index2 < guess.length(); index2++) {
          if(index != index2) {
             if(answer[index] == guess[index2]) {
               b++;
             }
          }
      }
    }

    return b;
}


bool isNumber(std::string &s) 
{
  for( char c : s )
  {
    if( !isdigit( c ) ) 
      return false;
  }

  return true;
}


int main(int argc, char *argv[])
{
    std::vector<std::string> resultset;

    for(int i = 0; i < 10; i++) {
       for(int j = 0; j < 10; j++) {
         for(int k = 0; k < 10; k++) {
           for(int m = 0; m < 10; m++) {
             if(i != j && i != k && i != m && j != k && j != m && k !=m) {
                 char buff[5];
                 sprintf(buff, "%d%d%d%d", i, j, k, m);
                 resultset.push_back(buff);
              }
           }
         }
       }
    }

    srand (time(NULL));
    int number = rand() % resultset.size();
    std::string answer = resultset[number];

    while(1) {
        std::string line;
        std::cout << "Please input a number to guess: ";
        std::getline(std::cin, line);
        if(line.length() != 4 || isNumber(line) == false) {
            std::cout << "It is not a valid answer..." << std::endl << std::endl;
            continue;
        }

        int a = getA(answer, line);
        int b = getB(answer, line);
        std::cout << "Result: A = " << a << ", B = " << b << std::endl << std::endl;

        if(a == 4 && b == 0) {
             break;
        }
    }

    return 0;
}

